server/channels/store/localcachelayer/role_layer.go

```diff
@@ -60,15 +62,30 @@ func (s LocalCacheRoleStore) GetByNames(names []string) ([]*model.Role, error) {
 	var foundRoles []*model.Role
 	var rolesToQuery []string
 
-	for _, roleName := range names {
+	var toPass []any
+	for i := 0; i < len(names); i++ {
 		var role *model.Role
-		if err := s.rootStore.doStandardReadCache(s.rootStore.roleCache, roleName, &role); err == nil {
-			foundRoles = append(foundRoles, role)
+		toPass = append(toPass, &role)
+	}
+	errs := s.rootStore.doMultiReadCache(s.rootStore.roleCache, names, toPass)
+	for i, err := range errs {
+		if err != nil {
+			if err != cache.ErrKeyNotFound {
+				s.rootStore.logger.Warn("Error in Rolestore.GetByNames: ", mlog.Err(err))
+			}
+			rolesToQuery = append(rolesToQuery, names[i])
 		} else {
-			rolesToQuery = append(rolesToQuery, roleName)
+			gotRole := *(toPass[i].(**model.Role))
+			if gotRole != nil {
+				foundRoles = append(foundRoles, gotRole)
+			}
 		}
 	}
 
+	if len(rolesToQuery) == 0 {
```

Comment:
This was a MASSIVE bug which I uncovered. Even after we got all roles from cache, we would make an empty call to the DB, get no results anyways, and then return the cached results. This would basically make the cache useless for this call. Making this change should have a big impact.

Actually, I will send this improvement separately. This is not related to Redis at all. And this is also why we keep seeing high number of store calls to the Rolelayer. Because the cache layer doesn't work at all!

Reply:
:man_facepalming: great find!

Reply:
Actually, it's not a huge problem .. because of 

```go
func (s *SqlRoleStore) GetByNames(names []string) ([]*model.Role, error) {
	if len(names) == 0 {
		return []*model.Role{}, nil
	}
...
```

So the problem is that while we are making the calls to the store layer, which is shooting up the Grafana charts, we aren't _really_ making a DB call. So while it's good to do that change, there's no urgency to send it separately. This can go as part of this PR itself.

Reply:
Good we had that failsafe :sweat_smile: